tuna/gen_fastdb.py

```diff
@@ -0,0 +1,298 @@
+                 strict=False,
+                 verbosity=0,
+                 tag=''):
+  """check finddb and raise erros or log warnings"""
+  if cols_with_conv_params is None:
+    cols_with_conv_params = _DEFAULT_COLS_WITH_CONV_PARAMS
+
+  # pylint: disable-next=invalid-name
+  NO_ENTRIES, INVALID_ENTRIES, OPENCL_ON, MULTIPLE_SESSIONS, DUPLICATE_ENTRIES = 0, 1, 2, 3, 7
+  issues = []
+
+  logging.log(f'checking {tag}finddb...', end_char='\r')
+
+  if len(finddb) == 0:
+    if strict:
+      raise AssertionError(f'{tag}finddb contains no entry')
+    logging.error(f'{tag}finddb contains no entry')
+    issues.append(NO_ENTRIES)
+
+  if (finddb['valid'] != 1).any():
+    if strict:
+      raise AssertionError(f'{tag}finddb contains invalid entries')
+    logging.warning(f'{tag}finddb contains invalid entries')
+    issues.append(INVALID_ENTRIES)
+
+  if finddb['opencl'].any():
+    if strict:
+      raise AssertionError(
+          f'there are kernels with opencl enabled in {tag}finddb')
+    logging.warning(f'there are kernels with opencl enabled in {tag}finddb')
+    issues.append(OPENCL_ON)
+
+  if not df_tools.is_col_unique(finddb, col_name='session'):
+    if strict:
+      raise AssertionError(
+          f'data from multiple tuning sessions present in the {tag}finddb')
+    logging.warning(
+        f'data from multiple tuning sessions present in the {tag}finddb')
+    issues.append(MULTIPLE_SESSIONS)
+
+  num_duplicates = finddb.duplicated(subset=cols_with_conv_params +
+                                     ['solver']).sum()
+  if num_duplicates > 0:
+    if verbosity >= 1:
+      logging.log(f'{num_duplicates} duplicates found! generating report...',
+                  end_char='\r')
+      duplicates = finddb[finddb.duplicated(subset=cols_with_conv_params +
+                                            ['solver'],
+                                            keep=False)]
+      duplicates = duplicates.sort_values(cols_with_conv_params +
+                                          ['solver', 'kernel_time'])
+      duplicates = duplicates.groupby(cols_with_conv_params + ['solver'])
+
+      duplicates_str = as_heading("Duplicates") + '\n'
+      for i, (_, df) in enumerate(duplicates):  # pylint: disable=invalid-name
+        for _, row in df.iterrows():
+          conv_params_str = "+ " if i % 2 == 0 else "- "
+          for colname, val in zip(cols_with_conv_params,
+                                  row[cols_with_conv_params]):
+            conv_params_str += f'{colname} {val}, '
+          duplicates_str += f"{conv_params_str}\n  solver: {row['solver']},  " +\
+                    f"kernel_time: {row['kernel_time']}\n\n"
+
+      logging.log(duplicates_str, silent=True)
+
+    if strict:
+      raise AssertionError(
+          f'{num_duplicates} duplicate entries detected in {tag}finddb')
+    logging.warning(
+        f'{num_duplicates} duplicate entries delected in {tag}finddb')
+    issues.append(DUPLICATE_ENTRIES)
```

Comment:
if this is an issue, it should be covered by a unique_index in the DB and not something we should be checking for in python. We can discuss this offline on how to add/modify this constraint.